vissl/engines/benchmark_suite_scheduler.py

```diff
@@ -0,0 +1,545 @@
+        # Count total number of evaluation jobs.
+        total_jobs = 0
+        for benchmarks in self.evaluation_results.values():
+            total_jobs += len(benchmarks)
+
+        return len(self.evaluation_jobs_finished) == total_jobs
+
+    def _evaluate_checkpoints(self):
+        for checkpoint_str, benchmarks in self.evaluation_results.items():
+            # TODO: Can we possible retrieve this from CheckpointWriter, to consolidate logic.
+            checkpoint_str = os.path.join(
+                self.training_config.CHECKPOINT.DIR, f"{ checkpoint_str }.torch"
+            )
+            if PathManager.exists(checkpoint_str):
+                self._evaluate_checkpoint(checkpoint_str, benchmarks)
+
+    def _evaluate_checkpoint(self, checkpoint_str, benchmarks):
+        for benchmark in benchmarks:
+            retry_job = self._retry_job(benchmark)
+            if benchmark["job_id"] and not retry_job:
+                continue
+
+            if retry_job:
+                self.evaluation_jobs_finished.remove(benchmark["job_id"])
+                # Log the job retry.
+                job_id, slurm_state = benchmark["job_id"], benchmark["slurm_state"]
+                print(f"Retrying job: { job_id } in state: { slurm_state }")
+
+            args, config = self._generate_config(benchmark["config_files"])
+            job = self._launch_slurm_job(args, config)
+
+            time.sleep(15)  # Wait for slurm job status to be reliably updated.
+
+            # Set checkpoint status
+            benchmark["job_id"] = job.job_id
+            benchmark["num_retries"] += 1
+            benchmark["slurm_log_dir"] = config.SLURM.LOG_FOLDER
+            benchmark["slurm_checkpoint_dir"] = config.CHECKPOINT.DIR
+            benchmark[
+                "weights_init_params_file"
+            ] = config.MODEL.WEIGHTS_INIT.PARAMS_FILE
+            benchmark["slurm_state"] = job.state
+
+            log = f"""
+                Launched Slurm Evaluation job.
+                job_id: { job.job_id }, num_retries: { benchmark["num_retries"] }
+                evaluation_name: { benchmark["evaluation_name"] }
+                checkpoint_str: { checkpoint_str }
+                state_prev: None, state_current: { job.state }
+            """
+
+            print(log)
+
+            # Save evaluation results to json file.
+            self.save_evaluation_results()
+
+    def _retry_job(self, benchmark):
+        return benchmark["job_id"] in self.retry_evaluation_job_ids or (
+            benchmark["slurm_state"] in _SLURM_JOB_TERMINAL_STATES
+            and benchmark["slurm_state"] != "COMPLETED"
+            and self.auto_retry_evaluations
+            and benchmark["num_retries"] < self.max_retries
+        )
+
+    @retry(n_tries=_NUM_SLURM_RETRIES)
+    def _launch_slurm_job(self, args, config):
+        return launch_distributed_on_slurm(engine_name=args.engine_name, cfg=config)
+
+    def _write_json_file(self, data, file_name):
```

Comment:
can we reuse this function https://github.com/facebookresearch/vissl/blob/master/vissl/utils/io.py#L68 instead?

Reply:
No we can't. I tried this initially, but this function assumes a list of dicts and writes out the json in newline_json_delimited form.